### ISSUE TITLE
[questions] about benchmarks and auth

### ISSUE THREAD

**Issue by trsh** [cutelyst/cutelyst#122]
Why there so many variations?:

cutelyst-thread-post
another cutelyst-thread-post

cutelyst-pf-postgres

cutelyst-nginx-postg
cutelyst-thread-mysq
cutelyst-thread-mysq

cutelyst-pf-mysql-ep

and What does it all mean? Especially the 'pf' thing.

Is an oauth2 type authorization possible (token exchange) ? As I also intend to authenticate from mobile app, and there for session will not help?

**Comment by dantti** [comment]
Because there are many variations possible for running cutelyst.
The **pf** stands for Pre-Fork, while **thread** is that instead of pre forked processes you have a bunch of threads, the name looks the same but there's also a variation with EPoll event loop and without, but that only shows up if you put your mouse over the test.

The nginx test uses uWSGI instead of cutelyst-wsgi.

The idea is to see which is the best option for runnig cutelyst, currently it is with cutelyst-wsgi (instead of uWSGI) in threaded mode with EPoll event loop (which is default now on v2).

This might help a bit: https://github.com/cutelyst/FrameworkBenchmarks/blob/master/frameworks/C%2B%2B/cutelyst/benchmark_config.json

OAuth2 is possible yes, but you don't have any plugin to help you for that yet (patches welcome!), being mobile or not using a cookie-session works either way.

You can get inspiration from various Perl modules like:
http://search.cpan.org/~edenc/CatalystX-OAuth2-0.001002/lib/Catalyst/Authentication/Credential/OAuth2.pm

Since I didn't have the need I didn't implement that yet.

**Comment by trsh** [comment]
> which is default now on v2

So I have max perf. by default?

> being mobile or not using a cookie-session works either way

What do you mean? How doe's it work without cookies? Or you mean it doesn't?

> Since I didn't have the need I didn't implement that yet

Maybe this can be a Feature Request? I don't think I'm skilled enough to make this happen :/

**Comment by dantti** [comment]
Yes, since on Linux our custom EPoll dispatcher is much faster than the Qt's GLib default one I made it the default on v2.

If your website is just going to authenticate a user with your DB then our Session + Authentication modules are enough. As I understand you need OAuth2 if you are going to use authentication offered by FaceBook or Google+ for example.

I'd prefer opening a specific issue for that, but I must say I'd quite busy ATM só I can't promise on any dates.

**Comment by dantti** [comment]
Well I do mobile apps and yes I do use cookies, the authentication is done using the Basic header method, in terms of security you put HTTPS. IMHO I'd only use OAuth2 to integrate with 3rd party authenticators.

**Comment by dantti** [comment]
And the Basic header authentication doesn't even need cookies as the user/password can be sent on all requests.

**Comment by trsh** [comment]
@dantti interesting. Will look into it

**Comment by trsh** [comment]
However sending user and pass over every time gives me bad gut feelings :D

**Comment by dantti** [comment]
Well as I said you can use a Session cookie with Basic auth, and what would be the difference to sending user/pass only once or on every request? To properly secure this data you need HTTPS.

**Comment by trsh** [comment]
Still there are some good point to switch to OAuth

http://apiux.com/2013/07/10/oauth-2-trumps-basic-authentication/
https://auth0.com/blog/cookies-vs-tokens-definitive-guide/
